parallel-install/pkg/engine/engine.go

```diff
@@ -64,59 +71,57 @@ type Installation interface {
 	//However, once the underlying parallel operations end, the cancel condition is detected and the return channel is closed.
 	//All remaining components are not processed then.
 	//
-	Uninstall(ctx context.Context) (<-chan components.KymaComponent, error)
+	Uninstall(ctx context.Context) (<-chan components.KymaComponent, <-chan error, error)
 }
 
-func (e *Engine) Deploy(ctx context.Context) (<-chan components.KymaComponent, error) {
-
-	cmps, err := e.componentsProvider.GetComponents()
-	if err != nil {
-		return nil, err
-	}
-
+func (e *Engine) Deploy(ctx context.Context) (<-chan components.KymaComponent, <-chan error, error) {
```

Comment:
Do we really need 2 different channels to send updates to the client? Or would it be possible to use the status channel error field as a replacement for the error channel

Reply:
Unfortunately, we have some rare cases where an error can occur which isn't related to a component, e.g. https://github.com/tobiscr/hydroform/blob/send-error-updates/parallel-install/pkg/engine/engine.go#L121  or https://github.com/tobiscr/hydroform/blob/send-error-updates/parallel-install/pkg/engine/engine.go#L89

Reply:
I still think it can be avoided, for each link you provided:
1. Since we are returning when the error happens and there is no concurrency at that point inside `run`, `run` can just return an error without a channel. Then `Deploy` just gets it as a normal return value.
2. Same is valid for `Deploy`. The only moment where an error is sent to that channel is before calling `run`. That part of the code can be moved outside of the goroutine (since it anyway happens before `run` it does not benefit from the concurrency. And then we will use the current return error that is not used at all in `Deploy` as well.

Reply:
Makes sense.. FIXED :)